solutions/platinum/ioi-05-mountain.mdx

```diff
@@ -0,0 +1,178 @@
+
+	/** @return result of joining two ranges together */
+	inline array<int, 3> comb(const array<int, 3> &a, const array<int, 3> &b) {
+		return {max(a[0], a[1] + b[0]), a[1] + b[1], a[2] + b[2]};
+	}
+
+	/** builds the segment tree nodes */
+	void build(int v, int l, int r) {
+		if (l == r) t[v] = {0, 0, vals[l + 1] - vals[l]};
+		else {
+			int m = (l + r) / 2;
+			build(2 * v, l, m);
+			build(2 * v + 1, m + 1, r);
+			t[v] = comb(t[2 * v], t[2 * v + 1]);
+		}
+	}
+
+	/** applies the update to node v and places the lazy update */
+	void apply(int v, int x) {
+		int full_sub = t[v][2] * x;
+		if (x >= 0) {
+			t[v] = {full_sub, full_sub, t[v][2]};
+		} else {
+			t[v] = {0, full_sub, t[v][2]};
+		}
+		lz[v] = x;
+	}
+
+	/** pushes down the lazy update on node v to node v's children */
+	void pushdown(int v, int l, int r) {
+		if (lz[v] != LZ_ID && l != r) {
+			for (int x : {2 * v, 2 * v + 1}) { apply(x, lz[v]); }
+		}
+		lz[v] = LZ_ID;
+	}
+
+	void range_set(int v, int tl, int tr, int ql, int qr, int amt) {
+		if (qr < tl || ql > tr) return;
+		if (ql <= tl && tr <= qr) {
+			apply(v, amt);
+		} else {
+			pushdown(v, tl, tr);
+			int m = (tl + tr) / 2;
+			range_set(2 * v, tl, m, ql, qr, amt);
+			range_set(2 * v + 1, m + 1, tr, ql, qr, amt);
+			t[v] = comb(t[2 * v], t[2 * v + 1]);
+		}
+	}
+
+	int walk_right(int v, int l, int r, array<int, 3> cur, int mx) {
+		if (l == r) {
+			int dx = mx - cur[1];
+			int div = t[v][1] / t[v][2];
+			return vals[l] - 1 + dx / div;
+		} else {
+			int m = (l + r) / 2;
+			pushdown(v, l, r);
+			array<int, 3> x = comb(cur, t[v * 2]);
+			return x[0] > mx ? walk_right(2 * v, l, m, cur, mx)
+			                 : walk_right(2 * v + 1, m + 1, r, x, mx);
+		}
+	}
+
+  public:
+	LazySegtree(const vector<int> &arr)
+	    : sz((int)arr.size() - 1), t(4 * sz, ID), lz(4 * sz, LZ_ID), vals(arr) {
+		build(1, 0, sz - 1);
+	}
+
+	/** sets the range [ql, qr] (coordinate compresed indexes) to amt */
+	void range_set(int ql, int qr, int amt) {
+		range_set(1, 0, sz - 1, ql, qr, amt);
+	}
+
+	/** @return the last location where the height <= mx */
+	int walk_right(int mx) { return walk_right(1, 0, sz - 1, ID, mx); }
+
+	/** @return largest prefix sum in the whole array */
+	int best_prefix() { return t[1][0]; }
+};
+
+int main() {
+	int n;
+	cin >> n;
+	vector<vector<int>> tests;
+	vector<int> vals = {0};
+	while (true) {
+		char type;
+		cin >> type;
+		if (type == 'I') {
+			int a, b, d;
+			cin >> a >> b >> d;
+			tests.push_back({0, a, b, d});
+			vals.push_back(a);
+			vals.push_back(b + 1);
+		} else if (type == 'Q') {
+			int h;
+			cin >> h;
+			tests.push_back({1, h});
+		} else {
+			break;
+		}
+	}
+	sort(begin(vals), end(vals));
+	vals.erase(unique(begin(vals), end(vals)), end(vals));
+	const auto id = [&](int x) -> int {
+		return lower_bound(begin(vals), end(vals), x) - begin(vals);
+	};
+
+	LazySegtree st(vals);
+	for (int i = 0; i < (int)tests.size(); i++) {
+		const vector<int> &cur = tests[i];
```

Comment:
why don't you just iterate with an iterator then?

Reply:
good point LOL lemme do that rq